lib/dice/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![cfg_attr(not(test), no_std)]
+
+use core::{mem, slice, str::FromStr};
+use hkdf::Hkdf;
+use hubpack::SerializedSize;
+use salty::constants::SECRETKEY_SEED_LENGTH;
+use serde::{Deserialize, Serialize};
+use sha3::Sha3_256;
+use zeroize::Zeroize;
+
+mod cert;
+pub use crate::cert::{AliasCert, CertError, DeviceIdCert};
+mod alias_cert_tmpl;
+mod deviceid_cert_tmpl;
+mod handoff;
+pub use crate::handoff::{AliasData, Handoff};
+
+pub const SEED_LENGTH: usize = SECRETKEY_SEED_LENGTH;
+pub const SN_LENGTH: usize = deviceid_cert_tmpl::SN_LENGTH;
+
+pub trait SeedBuf {
+    fn as_bytes(&self) -> &[u8; SEED_LENGTH];
+}
+
+#[derive(Zeroize)]
+#[zeroize(drop)]
+pub struct Cdi([u8; SEED_LENGTH]);
+
+impl SeedBuf for Cdi {
+    fn as_bytes(&self) -> &[u8; SEED_LENGTH] {
+        &self.0
+    }
+}
+
+impl Cdi {
+    const REG_ADDR_NONSEC: u32 = 0x40000900;
+
+    // Read the CDI from registers & return to the caller.
+    // Clear the CDI after it's consumed to prevent future access.
+    pub fn new() -> Option<Self> {
```

Comment:
This function has side effects that I wouldn't expect for new().  Namely, calling it multiple times is guaranteed to fail which is closer to a singleton than a normal struct.

Reply:
Good point. I've renamed this function to 'from_reg' to indicate the constructor is populating the Cdi from related registers. The method used to clear the DICE registers has been simplified significantly as well. Finally I've documented the side effects more clearly / explicitly.